vllm/distributed/device_communicators/custom_all_reduce_utils.py

```diff
@@ -0,0 +1,165 @@
+import json
+import os
+import sys
+from contextlib import contextmanager
+from typing import Callable, Dict, List, Optional
+
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+
+import vllm.envs as envs
+from vllm.distributed.parallel_state import get_cpu_world_group, get_local_rank
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+@contextmanager
+def mute_output():
+    with open(os.devnull, "w") as f:
+        sys.stderr = f
+        sys.stdout = f
+        yield
+
+
+def producer(i, CUDA_VISIBLE_DEVICES=None):
```

Comment:
nit: Let's make this function private?
```suggestion
def _producer(i, CUDA_VISIBLE_DEVICES=None):
```

Reply:
fixed in [770e508](https://github.com/vllm-project/vllm/pull/4992/commits/770e50804d04e7f35cb28d79b947f12c38ebcf07) . we only allow exporting one function now.